### PR TITLE
HHH-17246 - Guard against Sybase being configured for truncating trailing zeros

### PR DESCRIPTION
**Pull request by jrenaat**
----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

https://hibernate.atlassian.net/browse/HHH-17246

### CODE REVIEW

**Comment by gavinking** [comment]
@jrenaat It should be possible to detect that Sybase is configured like this, right? Just like we detect certain things about how MySQL or Oracle is configured, when we instantiate their dialects.

I think we should start logging warnings for such "bad" config, i.e. this thing, and also non-ANSI nulls.

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	public static final UUIDasVarBinaryJdbcType INSTANCE = new UUIDasVarBinaryJdbcType();
+
+	@Override
+	public int getDdlTypeCode() {
+		return Types.VARBINARY;
+	}
+
+	@Override
+	public int getDefaultSqlTypeCode() {
+		return VARBINARY_UUID;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java:33]:
```suggestion
		return UUID;
```

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * @author Jan Schatteman
+ */
+public class UUIDasVarBinaryJdbcType extends BinaryJdbcType {
+
+	public static final UUIDasVarBinaryJdbcType INSTANCE = new UUIDasVarBinaryJdbcType();
+
+	@Override
+	public int getDdlTypeCode() {
+		return Types.VARBINARY;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java:28]:
```suggestion
		return Types.BINARY;
```

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	public int getDefaultSqlTypeCode() {
+		return VARBINARY_UUID;
+	}
+
+	@Override
+	public <X> ValueExtractor<X> getExtractor( JavaType<X> javaType ) {
+		return new BasicExtractor<>( javaType, this ) {
+			@Override
+			protected X doExtract( ResultSet rs, int paramIndex, WrapperOptions options ) throws SQLException {
+				return javaType.wrap( Arrays.copyOf(rs.getBytes( paramIndex), 16), options );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java:41]:
```suggestion
				final byte[] bytes = rs.getBytes( paramIndex );
				return javaType.wrap( bytes == null || bytes.length == 16 ? bytes : Arrays.copyOf( bytes, 16 ), options );
```

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+			}
+
+			@Override
+			protected X doExtract( CallableStatement statement, int index, WrapperOptions options ) throws SQLException {
+				return javaType.wrap( Arrays.copyOf(statement.getBytes(index), 16), options );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java:46]:
```suggestion
				final byte[] bytes = statement.getBytes( index );
				return javaType.wrap( bytes == null || bytes.length == 16 ? bytes : Arrays.copyOf( bytes, 16 ), options );
```

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+			}
+
+			@Override
+			protected X doExtract( CallableStatement statement, String name, WrapperOptions options )
+					throws SQLException {
+				return javaType.wrap( Arrays.copyOf(statement.getBytes(name), 16), options );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java:52]:
```suggestion
				final byte[] bytes = statement.getBytes( name );
				return javaType.wrap( bytes == null || bytes.length == 16 ? bytes : Arrays.copyOf( bytes, 16 ), options );
```

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+import java.sql.SQLException;
+import java.sql.Types;
+import java.util.Arrays;
+
+import static org.hibernate.type.SqlTypes.VARBINARY_UUID;
+
+/**
+ * @author Jan Schatteman
+ */
+public class UUIDasVarBinaryJdbcType extends BinaryJdbcType {
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UUIDasVarBinaryJdbcType.java:22]:
```suggestion
public class UuidAsBinaryJdbcType extends BinaryJdbcType {
```

File: hibernate-core/src/main/java/org/hibernate/type/SqlTypes.java [hunk 7]
```diff
@@ -423,6 +423,13 @@ public class SqlTypes {
 	 */
 	public static final int UUID = 3000;
 
+	/**
+	 * A type code representing the generic SQL type {@code UUID} when it's being stored as a VARBINARY type.
+	 *
+	 * @see org.hibernate.type.descriptor.jdbc.UUIDasVarBinaryJdbcType
+	 */
+	public static final int VARBINARY_UUID = 3020;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/SqlTypes.java:431]: Remove, that isn't needed

File: hibernate-core/src/main/java/org/hibernate/dialect/SybaseASEDialect.java [hunk 8]
```diff
@@ -246,6 +248,7 @@ public void contributeTypes(TypeContributions typeContributions, ServiceRegistry
 		final JdbcTypeRegistry jdbcTypeRegistry = typeContributions.getTypeConfiguration()
 				.getJdbcTypeRegistry();
 		jdbcTypeRegistry.addDescriptor( Types.BOOLEAN, TinyIntJdbcType.INSTANCE );
+		jdbcTypeRegistry.addDescriptor( SqlTypes.VARBINARY_UUID, UUIDasVarBinaryJdbcType.INSTANCE );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/dialect/SybaseASEDialect.java:251]: Remove

File: hibernate-core/src/main/java/org/hibernate/boot/model/process/spi/MetadataBuildingProcess.java [hunk 9]
```diff
@@ -787,7 +787,7 @@ public void contributeType(CompositeUserType<?> type) {
 			);
 		}
 		else {
-			addFallbackIfNecessary( jdbcTypeRegistry, SqlTypes.UUID, SqlTypes.BINARY );
+			addFallbackIfNecessary( jdbcTypeRegistry, SqlTypes.UUID, SqlTypes.VARBINARY_UUID );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/boot/model/process/spi/MetadataBuildingProcess.java:790]:
```suggestion
			jdbcTypeRegistry.addDescriptorIfAbsent( UuidAsVarbinaryJdbcType.INSTANCE );
```

File: hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+	@BeforeEach
+	void setUp(SessionFactoryScope scope) {
+		this.uuid = scope.fromTransaction( session -> {
+			UUID uuid = UUID.randomUUID();
+			// Create a UUID with trailing zeros
+			while ( SafeRandomUUIDGenerator.isSafeUUID(uuid) ) {
+				uuid = UUID.randomUUID();
+			}
+			final Book book = new Book(uuid, "John Doe");
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java:45]: Just use a static UUID.

```suggestion
			final Book book = new Book( UUID.fromString("53886a8a-7082-4879-b430-25cb94415b00") , "John Doe");
```

File: hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+								ResultSet rs = st.executeQuery( "select id from Book" );
+								rs.next();
+								byte[] barr = rs.getBytes( 1 );
+								assertEquals( 15, barr.length );
+							}
+					);
+				}
+		);
+		scope.inTransaction(
+				session -> {
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java:77]:
```suggestion
		scope.inTransaction(
				session -> {
					assertEquals( 15, session.createNativeQuery( "select id from Book", byte[].class ).getSingleResult() );
```

File: hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+/**
+ * @author Jan Schatteman
+ */
+@RequiresDialect(value = SybaseASEDialect.class)
+@DomainModel(annotatedClasses = { SybaseASEUUIDTest.Book.class })
+@SessionFactory
+public class SybaseASEUUIDTest {
+
+	private UUID uuid;
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java:35]:
```suggestion
```

File: hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+		);
+		scope.inTransaction(
+				session -> {
+					Book b = session.createQuery( "from Book", Book.class ).getSingleResult();
+					UUID uuid = b.id;
+					ByteBuffer bb = ByteBuffer.allocate( 8 );
+					bb.putLong( uuid.getLeastSignificantBits() );
+					byte[] arr = bb.array();
+					assertEquals(8, arr.length);
+					assertEquals(this.uuid, uuid);
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java:84]:
```suggestion
					assertEquals(UUID.fromString("53886a8a-7082-4879-b430-25cb94415b00"), uuid);
```

**Review by beikov — CHANGES_REQUESTED** [review]

**Comment by jrenaat** [comment]
> @jrenaat It should be possible to detect that Sybase is configured like this, right? Just like we detect certain things about how MySQL or Oracle is configured, when we instantiate their dialects.
> 
> I think we should start logging warnings for such "bad" config, i.e. this thing, and also non-ANSI nulls.

I tried to find out how to do this, but failed miserably ...

File: hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+		scope.inTransaction(
+			session -> session.createMutationQuery( "delete from Book" ).executeUpdate()
+		);
+	}
+
+	@Test
+	@JiraKey( value = "HHH-17246" )
+	public void testTrailingZeroByteTruncation(SessionFactoryScope scope) {
+		scope.inSession(
+			session -> assertEquals( 15, session.createNativeQuery("select id from Book", byte[].class).getSingleResult().length )
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java:52]:
```suggestion
			session -> assertEquals( 16, session.createNativeQuery("select id from Book", byte[].class).getSingleResult().length )
```

File: hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+@SessionFactory
+public class SybaseASEUUIDTest {
+
+	private static final UUID uuid = UUID.fromString("53886a8a-7082-4879-b430-25cb94415b00");
+
+	@BeforeEach
+	void setUp(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			final Book book = new Book(uuid, "John Doe");
+			session.persist( book );
```
[the remark this answers is not in the document]
Comment (jrenaat) [reply hibernate-core/src/test/java/org/hibernate/orm/test/id/uuid/SybaseASEUUIDTest.java:37]: That said, the `UuidAsBinaryJdbcType` seems to cause issues for Oracle, irrespective of what ddl type code is used (happens with binary, as well as varbinary)

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UuidAsBinaryJdbcType.java [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * @author Jan Schatteman
+ */
+public class UuidAsBinaryJdbcType extends BinaryJdbcType {
+
+	public static final UuidAsBinaryJdbcType INSTANCE = new UuidAsBinaryJdbcType();
+
+	@Override
+	public int getDdlTypeCode() {
+		return Types.VARBINARY;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/jdbc/UuidAsBinaryJdbcType.java:28]: I understand that the bug only emerges when using VARBINARY, but the correct type here is BINARY:

```suggestion
		return Types.BINARY;
```

File: hibernate-core/src/main/java/org/hibernate/type/descriptor/java/UUIDJavaType.java [hunk 17]
```diff
@@ -41,8 +41,8 @@ public String toString(UUID value) {
 		return ToStringTransformer.INSTANCE.transform( value );
 	}
 
-	public UUID fromString(CharSequence string) {
-		return ToStringTransformer.INSTANCE.parse( string.toString() );
+	public UUID fromString(CharSequence string)  {
+		return '-' == string.charAt( 8 ) ? ToStringTransformer.INSTANCE.parse( string.toString() ) : NoDashesStringTransformer.INSTANCE.parse( string.toString() );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/type/descriptor/java/UUIDJavaType.java:45]: I don't mind this change, but is it necessary to get some test working?

Comment (jrenaat) [reply hibernate-core/src/main/java/org/hibernate/type/descriptor/java/UUIDJavaType.java:45]: Yes, the new UUID type is causing some issues with json in oracle; this changes fixes a number of tests in `JsonEmbeddableTest`, but there is still one remaining (`testSelectionItems`), that as far as i can tell has to with the function hexToRaw not liking that the UUID contain '-'.

**Pull request merged**